contracts/src/v0.8/dev/automation/2_1/AutomationRegistrar2_1.sol

```diff
@@ -0,0 +1,518 @@
+
+  event RegistrationApproved(bytes32 indexed hash, string displayName, uint256 indexed upkeepId);
+
+  event RegistrationRejected(bytes32 indexed hash);
+
+  event AutoApproveAllowedSenderSet(address indexed senderAddress, bool allowed);
+
+  event ConfigChanged(address keeperRegistry, uint96 minLINKJuels);
+
+  error InvalidAdminAddress();
+  error RequestNotFound();
+  error HashMismatch();
+  error OnlyAdminOrOwner();
+  error InsufficientPayment();
+  error RegistrationRequestFailed();
+  error OnlyLink();
+  error AmountMismatch();
+  error SenderMismatch();
+  error FunctionNotPermitted();
+  error LinkTransferFailed(address to);
+  error InvalidDataLength();
+
+  /*
+   * @param LINKAddress Address of Link token
+   * @param autoApproveType setting for auto-approve registrations
+   * @param autoApproveMaxAllowed max number of registrations that can be auto approved
+   * @param keeperRegistry keeper registry address
+   * @param minLINKJuels minimum LINK that new registrations should fund their upkeep with
+   */
+  constructor(
+    address LINKAddress,
+    address keeperRegistry,
+    uint96 minLINKJuels,
+    UpkeepTypeConfig[] memory upkeepConfigs
+  ) ConfirmedOwner(msg.sender) {
+    LINK = LinkTokenInterface(LINKAddress);
+    setRegistrationConfig(keeperRegistry, minLINKJuels);
+    for (uint256 idx = 0; idx < upkeepConfigs.length; idx++) {
+      setUpkeepTypeConfig(upkeepConfigs[idx]);
+    }
+  }
+
+  //EXTERNAL
+
+  /**
+   * @notice register can only be called through transferAndCall on LINK contract
+   * @param name string of the upkeep to be registered
+   * @param encryptedEmail email address of upkeep contact
+   * @param upkeepContract address to perform upkeep on
+   * @param gasLimit amount of gas to provide the target contract when performing upkeep
+   * @param adminAddress address to cancel upkeep and withdraw remaining funds
+   * @param triggerType the type of trigger for the upkeep
+   * @param checkData data passed to the contract when checking for upkeep
+   * @param triggerConfig the config for the trigger
+   * @param offchainConfig offchainConfig for upkeep in bytes
+   * @param amount quantity of LINK upkeep is funded with (specified in Juels)
+   * @param sender address of the sender making the request
+   */
+  function register(
+    string memory name,
+    bytes calldata encryptedEmail,
+    address upkeepContract,
+    uint32 gasLimit,
+    address adminAddress,
+    uint8 triggerType,
+    bytes memory checkData,
+    bytes memory triggerConfig,
+    bytes memory offchainConfig,
+    uint96 amount,
+    address sender
+  ) external onlyLINK {
+    _register(
+      RegistrationParams({
+        name: name,
+        encryptedEmail: encryptedEmail,
+        upkeepContract: upkeepContract,
+        gasLimit: gasLimit,
+        adminAddress: adminAddress,
+        triggerType: triggerType,
+        checkData: checkData,
+        triggerConfig: triggerConfig,
+        offchainConfig: offchainConfig,
```

Comment:
Are we still using this offchain config anymore if we are doing an explicit trigger config?

Reply:
technically we only need one arbitrary bytes field. 'offchainConfig' name might be better to extend to other use cases?

Reply:
ongoing debate. I think we could refactor into 1 field. It seems two only made more sense to me :sweat_smile:. Doesn't feel high priority to me though.